### PR TITLE
[ALLUXIO-2047] Surfacing non-Alluxio directories from object storage

### PR DESCRIPTION
**Pull request by apc999**
https://alluxio.atlassian.net/browse/ALLUXIO-2035
This is a continuation of PR #3615

### CODE REVIEW

**Comment by gpang** [comment]
Jenkins, test this please

File: underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java [hunk 1]
```diff
@@ -515,42 +516,44 @@ private boolean isRoot(String key) {
...
-      path = PathUtils.normalizePath(path, PATH_SEPARATOR);
-      path = path.equals(PATH_SEPARATOR) ? "" : path;
-      // Gets all the objects under the path, because we have no idea if there are non Alluxio
-      // managed "directories"
-      GSObject[] objs = mClient.listObjects(mBucketName, path, "");
-      if (recursive) {
-        String[] ret = new String[objs.length];
-        for (int i = 0; i < objs.length; i++) {
+      while (true) {
+        StorageObjectsChunk chunk = mClient.listObjectsChunked(mBucketName, path, delimiter, 1000L,
```
Comment (gpang) [inline underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java]: The `1000L` should be a constant?

File: underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java [hunk 2]
```diff
@@ -489,42 +490,38 @@ private boolean isRoot(String key) {
...
-          child = stripFolderSuffixIfPresent(child);
-          ret[i] = child;
-        }
-        return ret;
-      }
-
-      // Non recursive, so set the delimiter, let the listObjects only get the files in the folder
-      listObjectsRequest.setDelimiter(PATH_SEPARATOR);
-      Set<String> children = new HashSet<String>();
+      Set<String> children = new HashSet<>();
```
Comment (gpang) [inline underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java]: why is this a `Set` and not a `List` like the others? When might duplicate children happen?

File: underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java [hunk 3]
```diff
@@ -489,42 +490,38 @@ private boolean isRoot(String key) {
...
       ObjectListing listing = mClient.listObjects(listObjectsRequest);
       for (OSSObjectSummary objectSummary : listing.getObjectSummaries()) {
         // Remove parent portion of the key
         String child = getChildName(objectSummary.getKey(), path);
-        // Remove any portion after the path delimiter
-        int childNameIndex = child.indexOf(PATH_SEPARATOR);
-        child = childNameIndex != -1 ? child.substring(0, childNameIndex) : child;
+        if (!recursive) {
```
Comment (gpang) [inline underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java]: why does this have to be handled, when the delimiter already depends on `recursive`? (the other implementations don't do anything specific with `recursive` after `delimiter`)

File: underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java [hunk 4]
```diff
@@ -525,7 +524,7 @@ private boolean isFolder(String key) {
         S3Object[] objs = mClient.listObjects(mBucketName, path, "");
         // If there are, this is a folder and we can create the necessary metadata
         if (objs.length > 0) {
-          mkdirsInternal(path);
+          // TODO(binfan): Enable optimization "mkdirsInternal(path);"
```
Comment (gpang) [inline underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java]: why was this removed?

File: underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java [hunk 5]
```diff
@@ -557,47 +556,44 @@ private boolean isRoot(String key) {
...
-      path = PathUtils.normalizePath(path, PATH_SEPARATOR);
-      path = path.equals(PATH_SEPARATOR) ? "" : path;
-      // Gets all the objects under the path, because we have no idea if there are non Alluxio
-      // managed "directories"
-      S3Object[] objs = mClient.listObjects(mBucketName, path, "");
-      if (recursive) {
-        List<String> ret = new ArrayList<>();
-        for (S3Object obj : objs) {
+      while (true) {
+        StorageObjectsChunk chunk = mClient.listObjectsChunked(mBucketName, path, delimiter, 1000L,
```
Comment (gpang) [inline underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java]: The `1000L` should be a constant?

**Comment by gpang** [comment]
@apc999 Thanks for the fix! I left a few questions.

File: underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java (hunk 1 above)
Comment (apc999) [inline underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java]: done

File: underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java (hunk 2 above)
Comment (apc999) [inline underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java]: Good catch. Actually, the others should use `Set` too like this as duplicated children are possible.
Updated

File: underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java (hunk 3 above)
Comment (apc999) [inline underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java]: removed

File: underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java (hunk 5 above)
Comment (apc999) [inline underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java]: done

File: underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java (hunk 4 above)
Comment (apc999) [inline underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java]: this is an optimization to convert an S3 encoded dir into an Alluxio encoded dir, which requires write on reading a S3 bucket as ufs. This is introduced as an optimization after 1.1.0. After discussion with Jiri, we think this might have side-effect (e.g., fail to list on a read-only bucket), and let's hide it for this patch release.

File: underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java (hunk 2 above)
Comment (yupeng9) [inline underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java]: Can there be duplicated children? If so, then it's possible that this method returns less than 1k objects, but still hits the client limit.

Comment (apc999) [inline underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java]: Children will be unique. However, it is possible that the same dir have been encoded twice: once by Alluxio (`mkdirInternal`), once from the `commonPrefixes` returned by jets3t. See the incline comment I added from a real example run

File: underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java (hunk 4 above)
Comment (apc999) [inline underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java]: I reverted my removal as it will require more change.

File: underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java [hunk 6]
```diff
@@ -507,50 +512,71 @@ private boolean isRoot(String key) {
...
-      path = path.equals(PATH_SEPARATOR) ? "" : path;
-      // Gets all the objects under the path, because we have no idea if there are non Alluxio
-      // managed "directories"
-      GSObject[] objs = mClient.listObjects(mBucketName, path, "");
-      if (recursive) {
-        String[] ret = new String[objs.length];
-        for (int i = 0; i < objs.length; i++) {
+      boolean done = false;
+      while (!done) {
+        // Files/dirs in GCS UFS can be possibly encoded differently:
```
Comment (jsimsa) [inline underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java]: I would just focus on directories here:

```
Directories in GCS UFS can be possibly encoded in two different ways:
(1) as file objects with FOLDER_SUFFIX for directories created through Alluxio or
(2) as "common prefixes" of other files objects for directories not created through Alluxio

We iterate over chunk.getObjects to account for (1) and over chunk.getCommonPrefixes to account for (2).
...
```

File: underfs/gcs/src/test/java/alluxio/underfs/gcs/GCSUnderFileSystemTest.java [hunk 7]
```diff
@@ -14,13 +14,18 @@
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
-import org.powermock.api.mockito.PowerMockito;
+import org.junit.runner.RunWith;
```
Comment (jsimsa) [inline underfs/gcs/src/test/java/alluxio/underfs/gcs/GCSUnderFileSystemTest.java]: why are changes in this file needed?

Comment (apc999) [inline underfs/gcs/src/test/java/alluxio/underfs/gcs/GCSUnderFileSystemTest.java]: after marking `GCSUnderFileSystem` final

File: underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java [hunk 8]
```diff
@@ -477,7 +477,8 @@ private boolean isRoot(String key) {
 
   /**
    * Lists the files in the given path, the paths will be their logical names and not contain the
-   * folder suffix.
+   * folder suffix. Note that, due to the limitation of OSS client, this method can only return up
+   * to 1000 objects.
```
Comment (jsimsa) [inline underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java]: @yupeng9 could you confirm with Alibaba that there is no way to list more than 1000 objects? thanks

File: underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java (hunk 6 above)
Comment (apc999) [inline underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java]: thanks, will update it

File: underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java [hunk 9]
```diff
@@ -489,42 +490,33 @@ private boolean isRoot(String key) {
...
-      // Gets all the objects under the path, because we have no idea if there are non Alluxio
-      // managed "directories"
+      // If non-recursive, let the listObjects only get the files in the folder
+      String delimiter = recursive ? null : PATH_SEPARATOR;
+
+      // NOTE(binfan): currently OSS client only supports listing at most 1000 objects, without
+      // setting continuation
       ListObjectsRequest listObjectsRequest = new ListObjectsRequest(mBucketName);
       listObjectsRequest.setPrefix(path);
+      listObjectsRequest.setMaxKeys(1000);
```
Comment (jsimsa) [inline underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java]: define 1000 as a named constant `LISTING_LENGTH`

File: underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java [hunk 10]
```diff
@@ -489,42 +490,33 @@ private boolean isRoot(String key) {
       path = stripPrefixIfPresent(path);
       path = PathUtils.normalizePath(path, PATH_SEPARATOR);
       path = path.equals(PATH_SEPARATOR) ? "" : path;
-      // Gets all the objects under the path, because we have no idea if there are non Alluxio
-      // managed "directories"
+      // If non-recursive, let the listObjects only get the files in the folder
```
Comment (jsimsa) [inline underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java]: we should add the "(1), (2) ..." comment you have in GCS here as well

File: underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java [hunk 11]
```diff
@@ -549,55 +552,70 @@ private boolean isRoot(String key) {
...
-      path = path.equals(PATH_SEPARATOR) ? "" : path;
-      // Gets all the objects under the path, because we have no idea if there are non Alluxio
-      // managed "directories"
-      S3Object[] objs = mClient.listObjects(mBucketName, path, "");
-      if (recursive) {
-        List<String> ret = new ArrayList<>();
-        for (S3Object obj : objs) {
+      boolean done = false;
+      while (!done) {
+        // Files/dirs in S3 UFS can be possibly encoded differently:
```
Comment (jsimsa) [inline underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java]: similar comment here as for GCS (focusing on directories)

File: underfs/s3/src/test/java/alluxio/underfs/s3/S3UnderFileSystemTest.java [hunk 12]
```diff
@@ -14,22 +14,27 @@
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
-import org.powermock.api.mockito.PowerMockito;
+import org.junit.runner.RunWith;
```
Comment (jsimsa) [inline underfs/s3/src/test/java/alluxio/underfs/s3/S3UnderFileSystemTest.java]: why are changes in this file needed?

File: underfs/s3/src/test/java/alluxio/underfs/s3/S3UnderFileSystemTest.java [hunk 13]
```diff
@@ -14,22 +14,27 @@
...
 import org.powermock.reflect.Whitebox;
 
 /**
  * Tests for the private helper methods in {@link S3UnderFileSystem} that do not require an S3
  * backend.
  */
-public class S3UnderFileSystemTest {
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(S3UnderFileSystem.class)
+public final class S3UnderFileSystemTest {
```
Comment (jsimsa) [inline underfs/s3/src/test/java/alluxio/underfs/s3/S3UnderFileSystemTest.java]: would it be difficult to add a test that checks that the chunked listing works (e.g. by setting LISTING_LENGTH to 1 and trying to list a couple on files)

**Comment by jsimsa** [comment]
@apc999 thank you for this improvement, left some comments

File: underfs/gcs/src/test/java/alluxio/underfs/gcs/GCSUnderFileSystemTest.java (hunk 7 above)
Comment (jsimsa) [inline underfs/gcs/src/test/java/alluxio/underfs/gcs/GCSUnderFileSystemTest.java]: ack

File: underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java [hunk 14]
```diff
@@ -549,55 +552,70 @@ private boolean isRoot(String key) {
...
    * @return an array of the file and folder names in this directory
    * @throws IOException if an I/O error occurs
    */
   private String[] listInternal(String path, boolean recursive) throws IOException {
+    path = stripPrefixIfPresent(path);
+    path = PathUtils.normalizePath(path, PATH_SEPARATOR);
+    path = path.equals(PATH_SEPARATOR) ? "" : path;
+    String delimiter = recursive ? "" : PATH_SEPARATOR;
+    String priorLastKey = null;
+    Set<String> listResult = new HashSet<>();
```
Comment (gpang) [inline underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java]: NIT: no longer should be called listResult.

File: underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java [hunk 15]
```diff
@@ -507,50 +512,71 @@ private boolean isRoot(String key) {
...
    * @return an array of the file and folder names in this directory
    * @throws IOException if an I/O error occurs
    */
   private String[] listInternal(String path, boolean recursive) throws IOException {
+    path = stripPrefixIfPresent(path);
+    path = PathUtils.normalizePath(path, PATH_SEPARATOR);
+    path = path.equals(PATH_SEPARATOR) ? "" : path;
+    String delimiter = recursive ? "" : PATH_SEPARATOR;
+    String priorLastKey = null;
+    Set<String> listResult = new HashSet<>();
```
Comment (gpang) [inline underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java]: NIT: no longer should be called listResult.

**Comment by gpang** [comment]
@apc999 Thanks for the fixes! I left a few minor comments.

File: underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java (hunk 9 above)
Comment (apc999) [inline underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java]: done

File: underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java (hunk 15 above)
Comment (apc999) [inline underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java]: renamed to children

File: underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java (hunk 6 above)
Comment (apc999) [inline underfs/gcs/src/main/java/alluxio/underfs/gcs/GCSUnderFileSystem.java]: done

File: underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java (hunk 10 above)
Comment (apc999) [inline underfs/oss/src/main/java/alluxio/underfs/oss/OSSUnderFileSystem.java]: done

File: underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java (hunk 14 above)
Comment (apc999) [inline underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java]: renamed to `children`

File: underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java (hunk 11 above)
Comment (apc999) [inline underfs/s3/src/main/java/alluxio/underfs/s3/S3UnderFileSystem.java]: done

File: underfs/s3/src/test/java/alluxio/underfs/s3/S3UnderFileSystemTest.java (hunk 13 above)
Comment (apc999) [inline underfs/s3/src/test/java/alluxio/underfs/s3/S3UnderFileSystemTest.java]: there is right now no harness to mock s3 (or GCS, OSS) client and test this. It requires a lot of new code. Not sure if I should make it in this PR as this is 1.1.1 blocking.
On the other hand, current tests coverage of ufs module is pretty low. I will create a ticket on this.

File: underfs/gcs/src/test/java/alluxio/underfs/gcs/GCSUnderFileSystemTest.java [hunk 16]
```diff
@@ -14,13 +14,18 @@
...
+import org.powermock.modules.junit4.PowerMockRunner;
 import org.powermock.reflect.Whitebox;
 
 /**
  * Tests for the private helper methods in {@link GCSUnderFileSystem} that do not require an GCS
  * backend.
  */
+@RunWith(PowerMockRunner.class)
+@PrepareForTest(GCSUnderFileSystem.class)
 public final class GCSUnderFileSystemTest {
```
Comment (apc999) [inline underfs/gcs/src/test/java/alluxio/underfs/gcs/GCSUnderFileSystemTest.java]: replied in the later comment.

**Comment by jsimsa** [comment]
LGTM
